ta/pkcs11/src/processing_rsa.c

```diff
@@ -0,0 +1,173 @@
+#include <tee_internal_api.h>
+#include <tee_internal_api_extensions.h>
+
+#include "attributes.h"
+#include "processing.h"
+
+static enum pkcs11_rc tee2pkcs_rsa_attributes(struct obj_attrs **pub_head,
+					      struct obj_attrs **priv_head,
+					      TEE_ObjectHandle tee_obj)
+{
+	enum pkcs11_rc rc = PKCS11_CKR_GENERAL_ERROR;
+	void *a_ptr = NULL;
+
+	rc = tee2pkcs_add_attribute(pub_head, PKCS11_CKA_MODULUS, tee_obj,
+				    TEE_ATTR_RSA_MODULUS);
+	if (rc)
+		goto out;
+
+	rc = get_attribute_ptr(*pub_head, PKCS11_CKA_PUBLIC_EXPONENT, &a_ptr,
+			       NULL);
+	if (rc != PKCS11_CKR_OK && rc != PKCS11_RV_NOT_FOUND)
+		goto out;
+
+	if (rc == PKCS11_RV_NOT_FOUND || !a_ptr) {
+		rc = tee2pkcs_add_attribute(pub_head,
+					    PKCS11_CKA_PUBLIC_EXPONENT,
+					    tee_obj,
+					    TEE_ATTR_RSA_PUBLIC_EXPONENT);
+		if (rc)
+			goto out;
+	}
+
+	rc = tee2pkcs_add_attribute(priv_head, PKCS11_CKA_MODULUS, tee_obj,
+				    TEE_ATTR_RSA_MODULUS);
+	if (rc)
+		goto out;
+
+	rc = tee2pkcs_add_attribute(priv_head, PKCS11_CKA_PUBLIC_EXPONENT,
+				    tee_obj, TEE_ATTR_RSA_PUBLIC_EXPONENT);
+	if (rc)
+		goto out;
+
+	rc = tee2pkcs_add_attribute(priv_head, PKCS11_CKA_PRIVATE_EXPONENT,
+				    tee_obj, TEE_ATTR_RSA_PRIVATE_EXPONENT);
+	if (rc)
+		goto out;
+
+	rc = tee2pkcs_add_attribute(priv_head, PKCS11_CKA_PRIME_1, tee_obj,
+				    TEE_ATTR_RSA_PRIME1);
+	if (rc)
+		goto out;
+
+	rc = tee2pkcs_add_attribute(priv_head, PKCS11_CKA_PRIME_2, tee_obj,
+				    TEE_ATTR_RSA_PRIME2);
+	if (rc)
+		goto out;
+
+	rc = tee2pkcs_add_attribute(priv_head, PKCS11_CKA_EXPONENT_1, tee_obj,
+				    TEE_ATTR_RSA_EXPONENT1);
+	if (rc)
+		goto out;
+
+	rc = tee2pkcs_add_attribute(priv_head, PKCS11_CKA_EXPONENT_2, tee_obj,
+				    TEE_ATTR_RSA_EXPONENT2);
+	if (rc)
+		goto out;
+
+	rc = tee2pkcs_add_attribute(priv_head, PKCS11_CKA_COEFFICIENT, tee_obj,
+				    TEE_ATTR_RSA_COEFFICIENT);
+out:
+	return rc;
+}
+
+enum pkcs11_rc generate_rsa_keys(struct pkcs11_attribute_head *proc_params,
+				 struct obj_attrs **pub_head,
+				 struct obj_attrs **priv_head)
+{
+	enum pkcs11_rc rc = PKCS11_CKR_GENERAL_ERROR;
+	void *a_ptr = NULL;
+	void *public_exponent = NULL;
+	uint32_t a_size = 0;
+	TEE_ObjectHandle tee_obj = TEE_HANDLE_NULL;
+	TEE_Result res = TEE_ERROR_GENERIC;
+	uint32_t tee_size = 0;
+	TEE_Attribute tee_attrs[1] = { };
+	uint32_t tee_count = 0;
+
+	if (!proc_params || !*pub_head || !*priv_head)
+		return PKCS11_CKR_TEMPLATE_INCONSISTENT;
+
+	rc = get_attribute_ptr(*pub_head, PKCS11_CKA_MODULUS_BITS, &a_ptr,
+			       &a_size);
+	if (rc != PKCS11_CKR_OK || a_size != sizeof(uint32_t))
+		return PKCS11_CKR_TEMPLATE_INCONSISTENT;
+
+	TEE_MemMove(&tee_size, a_ptr, sizeof(uint32_t));
+
+	rc = get_attribute_ptr(*pub_head, PKCS11_CKA_PUBLIC_EXPONENT, &a_ptr,
+			       &a_size);
+	if (rc != PKCS11_CKR_OK && rc != PKCS11_RV_NOT_FOUND)
+		return rc;
+
+	if (rc == PKCS11_CKR_OK && a_ptr) {
+		public_exponent = TEE_Malloc(a_size, TEE_MALLOC_FILL_ZERO);
```

Comment:
why allocating a buffer. Can't we use `a_ptr` straight in `TEE_InitRefAttribute()` at line 119?

Reply:
That fixed something at some point of time related to re-use of same pointer or buffer. But now that I am looking at it I don't spot what it would have been. Should be safe to use reference as user for the memory is in same scope. And there are no mutators for `pub_head` in between. I'll run my battery of extended tests to be sure and if nothing breaks then I'll do the change.

Reply:
Removed buffer allocation.